src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -238,7 +239,21 @@ public void authorize(ClientState state) throws InvalidRequestException, Unautho
         for (Function function : getFunctions())
             state.ensurePermission(Permission.EXECUTE, function);
 
-        unmask = state.hasUnmaskPermission(table);
+        unmask = state.hasTablePermission(table, Permission.UNMASK);
+
+        if (!unmask && !state.hasTablePermission(table, Permission.SELECT_MASKED))
+        {
+            List<ColumnMetadata> queriedMaskedColumns = table.columns()
+                                                             .stream()
+                                                             .filter(ColumnMetadata::isMasked)
```

Comment:
Would it be better to filter `isRestricted` first?

Reply:
[`StatementRestrictions#isRestricted`](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java#L341-L346) involves building a collection [in most cases](https://github.com/apache/cassandra/blob/cassandra-4.1.0/src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java#L60) and then running a `List#contains`. In contrast, {{ColumnMetadata#isMasked}} just returns a boolean. So I think `isMasked` is faster than is `isRestricted`.

Reply:
sgtm